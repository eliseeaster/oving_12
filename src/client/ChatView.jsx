import React, { useEffect, useState } from "react";

export function ChatView({ chatPreview, username, messageApi }) {
  const [chatLog, setChatlog] = useState([]);
  const [message, setMessage] = useState("");
  const [ws, setWs] = useState();

  useEffect(() => {
    const ws = new WebSocket("ws://localhost:3000");
    ws.onmessage = (event) => {
      const { message, id, username } = JSON.parse(event.data);
      setChatlog((chatLog) => [...chatLog, { message, id, username }]);
    };
    ws.onopen = (event) => {
      ws.send(
        JSON.stringify({
          type: "login",
          username,
        })
      );
    };
    setWs(ws);
  }, []);

  function handleSubmitMessage(e) {
    e.preventDefault();
    if (ws.readyState !== 0)
      ws.send(
        JSON.stringify({
          type: "message",
          message: message,
        })
      );
    setMessage("");
    console.log("message: " + message);
  }

  async function submit(e) {
    e.preventDefault();
    await messageApi.createMessage({ username, message });
  }

  const chatOutput = chatLog.length ? chatLog : chatPreview;
  return (
    <div id="chatContainer">
      <header>
        <h1>Chat page</h1>
      </header>
      <main id="chatLog">
        {chatOutput?.map(({ message, id, username }) => (
          <div className="message" keys={id}>
            <strong>{username}</strong>
            {message}
          </div>
        ))}
      </main>
      <div>
        <form onSubmit={handleSubmitMessage}>
          <input
            type="text"
            autoFocus={true}
            value={message}
            onChange={(e) => setMessage(e.target.value)}
          />
          <button>Send</button>
          <button onClick={submit}>Save</button>
        </form>
      </div>
    </div>
  );
}
