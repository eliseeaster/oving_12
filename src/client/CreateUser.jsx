import React, { useState } from "react";
import { InputField } from "./inputField";
import { useHistory } from "react-router";

export function CreateUser({ userApi }) {
  const [name, setName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");

  const history = useHistory();

  async function submit(e) {
    e.preventDefault();
    await userApi.createUser({ name, lastName, email });
  }

  return (
    <form onSubmit={submit}>
      <h1>Create new user</h1>
      <InputField label={"Name"} value={name} onChangeValue={setName} />
      <InputField
        label={"Last name"}
        value={lastName}
        onChangeValue={setLastName}
      />
      <InputField label={"Email"} value={email} onChangeValue={setEmail} />
      <button>Submit</button>
    </form>
  );
}
