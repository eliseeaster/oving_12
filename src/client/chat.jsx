import React, { useEffect, useState } from "react";
import "./index.jsx";

export function ChatPage({ messageApi }) {
  const [username, setUsername] = useState();
  if (!username) {
    return <ChatLogin onLogin={(username) => setUsername(username)} />;
  }
  return <ChatView username={username} messageApi={messageApi} />;
}

function ChatLogin({ onLogin }) {
  const [username, setUsername] = useState("");
  function handleSubmit(e) {
    e.preventDefault();
    onLogin(username);
    console.log("username" + username);
  }
  return (
    <div>
      <h1>Please log in</h1>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          value={username}
          onChange={(e) => setUsername(e.target.value)}
        />
        <button>Login</button>
      </form>
    </div>
  );
}

export function ChatView({ username, messageApi }) {
  const [chatLog, setChatlog] = useState([]);
  const [message, setMessage] = useState("");
  const [ws, setWs] = useState();

  useEffect(() => {
    const ws = new WebSocket("ws://localhost:3000");
    ws.onmessage = (event) => {
      const { message, id, username } = JSON.parse(event.data);
      setChatlog((chatLog) => [...chatLog, { message, id, username }]);
    };
    ws.onopen = (event) => {
      ws.send(
        JSON.stringify({
          type: "login",
          username,
        })
      );
    };
    setWs(ws);
  }, []);

  function handleSubmitMessage(e) {
    e.preventDefault();
    ws.send(
      JSON.stringify({
        type: "message",
        message: message,
      })
    );
    setMessage("");
    console.log("message: " + message);
  }

  async function submit(e) {
    e.preventDefault();
    console.log("name " + username + "message " + message);
    console.log("api " + messageApi);
    await messageApi.createMessage({ username, message });
  }

  return (
    <>
      <h1>Chat page</h1>
      <div>
        {chatLog.map(({ message, id, username }) => (
          <div keys={id}>
            <strong>{username}</strong>
            {message}
          </div>
        ))}
      </div>
      <div>
        <form onSubmit={handleSubmitMessage}>
          <input
            type="text"
            autoFocus={true}
            value={message}
            onChange={(e) => setMessage(e.target.value)}
          />
          <button>Send</button>
          <button onClick={submit}>Save</button>
        </form>
      </div>
    </>
  );
}
export default ChatPage;
