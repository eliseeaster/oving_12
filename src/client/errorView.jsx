import * as React from "react";
import { Link } from "react-router-dom";

export function ErrorView({ error }) {
  if (error.status === 401) {
    return (
      <div>
        You are not logged in{" "}
        <Link to={"/home"}>
          <button>Log in</button>
        </Link>
      </div>
    );
  }
  return (
    <>
      <div>Something went wrong: {error.toString()}</div>
    </>
  );
}
