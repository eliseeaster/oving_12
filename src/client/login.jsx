import React, { useState } from "react";
import "./index.jsx";
import { fetchJSON } from "./http";

export function randomString(length) {
  const possible =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmopqrstuvwxyz1234567890";
  let result = "";
  for (let i = 0; i < length; i++) {
    result += possible.charAt(Math.floor(Math.random() * possible.length));
  }
  return result;
}

export function Login({ identityProvider }) {
  const { discoveryURL, client_id } = identityProvider;

  async function handleLogin() {
    const { authorization_endpoint } = await fetchJSON(discoveryURL);

    const state = randomString(30);
    const loginState = { state };
    sessionStorage.setItem("loginState", JSON.stringify(loginState));

    const params = {
      client_id,
      response_type: "token",
      scope: "openid email profile",
      redirect_uri: window.location.origin + "/login/callback",
      state,
    };
    window.location.href =
      authorization_endpoint + "?" + new URLSearchParams(params);
  }

  return (
    <div className="loginDiv">
      <h1 className="loginTxt">Login</h1>
      <button className="loginBtn" onClick={handleLogin}>
        Login
      </button>
    </div>
  );
}

export default Login;
