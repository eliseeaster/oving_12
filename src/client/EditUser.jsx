import React, { useState } from "react";
import { useHistory, useParams } from "react-router";
import { InputField } from "./inputField";
import { useLoading } from "./useLoading";
import { ErrorView } from "./errorView";
import { LoadingView } from "./loadingView";

function EditUserForm({ user, onSubmit }) {
  const [name, setName] = useState(user.name);
  const [lastName, setLastName] = useState(user.lastName);
  const [email, setEmail] = useState(user.email);

  async function submit(e) {
    e.preventDefault();
    onSubmit(e, { name, lastName, email });
  }

  return (
    <form onSubmit={submit}>
      <h1>Edit an existing user ({name})</h1>
      <InputField label={"Name"} value={name} onChangeValue={setName} />
      <InputField
        label={"Last name"}
        value={lastName}
        onChangeValue={setLastName}
      />
      <InputField label={"Email"} value={email} onChangeValue={setEmail} />
      <button>Submit</button>
    </form>
  );
}

export function EditUser({ userApi }) {
  const { id } = useParams();
  const history = useHistory;

  const { data: user, loading, error, reload } = useLoading(
    async () => await userApi.getUser(id),
    [id]
  );

  async function handleSubmit(e, { name, lastName, email }) {
    e.preventDefault();
    await userApi.updateUser(id, { name, lastName, email });
    history.push("/");
  }

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading || !user) {
    return <LoadingView />;
  }

  return <EditUserForm user={user} onSubmit={handleSubmit} />;
}
