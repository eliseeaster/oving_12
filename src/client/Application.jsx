import React, { useState } from "react";
import { Link, Redirect, Route, Switch } from "react-router-dom";
import { UserInfo } from "./userInfo";
import { CreateUser } from "./createUser";
import Login from "./login";
import { ChatPage } from "./chat";
import { EditUser } from "./EditUser";
import { fetchJSON, postJSON } from "./http";
import { LoginCallBackPage } from "./LoginCallBackPage";
import { LoginInfoPage } from "./LoginInfoPage";
import { Messages } from "./messages";

export function Application() {
  const userApi = {
    listUsers: async () => await fetchJSON("/api/users"),
    getUser: async (id) => await fetchJSON(`/api/users/${id}`),
    createUser: async ({ name, lastName, email }) => {
      return postJSON("http://localhost:3000/api/users", {
        method: "POST",
        json: { name, lastName, email },
      });
    },
    updateUser: async (id, { name, lastName, email }) =>
      postJSON(`/api/users/${id}`, {
        method: "PUT",
        json: { name, lastName, email },
      }),
    loadProfile: async () =>
      await fetchJSON("/api/login-info", {
        headers: {
          ...(access_token ? { Authorization: `Bearer ${access_token}` } : {}),
        },
      }),
  };

  const messageApi = {
    listMessages: async () => await fetchJSON("/api/messages"),
    getMessage: async (id) => await fetchJSON(`/api/messages/${id}`),
    createMessage: async ({ username, message }) => {
      return postJSON("/api/messages", {
        method: "POST",
        json: { username, message },
      });
    },
  };

  const [access_token, setAccess_token] = useState();
  const googleIdentityProvider = {
    discoveryURL:
      "https://accounts.google.com/.well-known/openid-configuration",
    client_id:
      "869093563251-h9ra6bludm19okni7jgaool77vq8773v.apps.googleusercontent.com",
  };

  return (
    <>
      <main>
        <Switch>
          <Route path="/create">
            <nav>
              <Link to={"/home"}>Back</Link>
            </nav>
            {!access_token ? (
              <Redirect to={"/"} />
            ) : (
              <CreateUser userApi={userApi} />
            )}
          </Route>

          <Route exact path="/users">
            <nav>
              <Link to={"/home"}>Back</Link>
            </nav>
            {!access_token ? (
              <Redirect to={"/"} />
            ) : (
              <UserInfo userApi={userApi} />
            )}
          </Route>

          <Route path="/chat">
            <nav>
              <Link to={"/home"}>Back</Link>
            </nav>
            {!access_token ? (
              <Redirect to={"/"} />
            ) : (
              <ChatPage messageApi={messageApi}></ChatPage>
            )}
          </Route>

          <Route path="/messages">
            {!access_token ? (
              <Redirect to={"/"} />
            ) : (
              <Messages messageApi={messageApi}></Messages>
            )}
          </Route>

          <Route path={"/"} exact>
            <Login identityProvider={googleIdentityProvider} />
          </Route>

          <Route path="/login/callback">
            <LoginCallBackPage
              identityProvider={googleIdentityProvider}
              onAccessToken={(access_token) => setAccess_token(access_token)}
            />
          </Route>

          <Route path="/login-info">
            <nav>
              <Link to={"/home"}>Back</Link>
            </nav>
            <LoginInfoPage userApi={userApi} />
          </Route>

          <Route path="/users/:id/edit">
            <nav>
              <Link to={"/home"}>Back</Link>
            </nav>
            {!access_token ? (
              <Redirect to={"/"} />
            ) : (
              <EditUser userApi={userApi} />
            )}
          </Route>

          <Route exact path="/home">
            <h1>Welcome</h1>
            <div className="navContainer">
              <div>
                <Link to="/users">See users</Link>
              </div>
              <div>
                <Link to="/login-info">Profile</Link>
              </div>

              <div>
                <Link to="/chat">Chat</Link>
              </div>

              <div>
                <Link to="/messages">Messages</Link>
              </div>

              <div>
                <Link to="/create">Create</Link>
              </div>
            </div>
          </Route>
          <Route>Not found</Route>
        </Switch>
      </main>
    </>
  );
}
