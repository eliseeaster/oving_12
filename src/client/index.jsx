import ReactDOM from "react-dom";
import React from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import { fetchJSON, postJSON } from "./http";
import { Application } from "./Application";

function App() {
  const userApi = {
    listUsers: async () => await fetchJSON("/api/users"),
    getUser: async (id) => await fetchJSON(`/api/users/${id}`),
    createUser: async ({ name, lastName, email }) => {
      return postJSON("/api/users", {
        method: "POST",
        json: { name, lastName, email },
      });
    },
    updateUser: async (id, { name, lastName, email }) =>
      postJSON(`/api/users/${id}`, {
        method: "PUT",
        json: { name, lastName, email },
      }),
  };

  return (
    <Router>
      <Application userApi={userApi} />
    </Router>
  );
}

ReactDOM.render(<App />, document.getElementById("root"));
