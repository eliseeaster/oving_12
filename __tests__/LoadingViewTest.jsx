import * as React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { LoadingView } from "../src/client/loadingView";

describe("loading view", () => {
  it("shows loading view on dom", () => {
    const container = document.createElement("div");
    document.body.appendChild(container);
    act(() => {
      ReactDOM.render(<LoadingView />, container);
    });
    expect(container.innerHTML).toMatchSnapshot();
  });
});
