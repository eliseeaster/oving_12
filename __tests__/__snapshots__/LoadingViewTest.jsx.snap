// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading view shows loading view on dom 1`] = `"<h1>Loading...</h1>"`;
